### PR TITLE
Custom collection/global endpoints

### PR DESCRIPTION
**Pull request by GeorgeyB**
## Description

This PR introduces support for the `endpoints` configuration property for both collections and globals. Custom endpoints can be defined as described by [@jmikrut](https://github.com/payloadcms/payload/discussions/587#discussioncomment-2827918) in the corresponding discussion.

Endpoints are compared against succeeding entity endpoints using `path-to-regexp` (the same library that express uses). Any conflicts will be identified and flagged during validation with descriptive messages. This includes default endpoints set out by Payload, which are described in the same way in collection/global "defaults" files.

Endpoints are mounted to the appropriate router using mountEndpoints - note that because we now have default Payload endpoints described in defaults, we can use those defaults as a source of truth and mount them in the same way.

A new "Endpoints" collection has been created in the demo configuration to test this new feature.

- [x] I have read and understand the CONTRIBUTING.md document in this repository

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Existing test suite passes locally with my changes
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: src/init/mountEndpoints.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { Router } from 'express';
+
+import { Payload } from '../index';
+import { Endpoint } from '../config/types';
+
+function mountEndpoints(router: Router, endpoints: Endpoint[], ctx: Payload): void {
+  endpoints.forEach((endpoint) => {
+    endpoint.handlers.forEach((handler) => {
```
Comment (jmikrut) [inline src/init/mountEndpoints.ts:8]: To simplify, can we just do `router[endpoint.method](endpoint.route, endpoints)`?

File: src/init/mountEndpoints.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+      router[endpoint.method].call(router, endpoint.route, handler.bind(ctx));
```
Comment (jmikrut) [inline src/init/mountEndpoints.ts:9]: I think you are binding `this` to expose Payload within these handlers. But, Payload is _**already available**_ and bound to `req` at this stage via Payload middleware further up the stack (can access via `req.payload`). So, binding `this` is not necessary IMO.

**Review by jmikrut — CHANGES_REQUESTED** [review]
Everything looks great! I believe @DanRibbens is working on documentation here and I have a few changes—namely just removing `this` binding as it is not necessary.

File: src/init/mountEndpoints.ts (hunk 2 above)
Comment (DanRibbens) [reply src/init/mountEndpoints.ts:9]: > Everything looks great! I believe @DanRibbens is working on documentation here and I have a few changes—namely just removing `this` binding as it is not necessary.

I've already made this change and am finishing up some other improvements. @GeorgeyB standby, I'm going to push soon.

File: src/init/mountEndpoints.ts (hunk 1 above)
Comment (GeorgeyB) [reply src/init/mountEndpoints.ts:8]: Sure! `.call` is merely personal preference due to readability.

File: src/init/mountEndpoints.ts (hunk 2 above)
Comment (GeorgeyB) [reply src/init/mountEndpoints.ts:9]: @jmikrut This was to maintain consistency with the Payload's existing handlers where `this` is bound to the Payload context, but you're absolutely right, it's unnecessary if it's already accessible in `req`.

@DanRibbens Nice.

Comment (DanRibbens) [reply src/init/mountEndpoints.ts:9]: I ran out of time working through these changes. I'll try and pick it up over the weekend to wrap this one up.

**Review by Dekbre — APPROVED** [review]

**Review by sameerayaan — APPROVED** [review]
**_[]()_**

**Pull request merged**